Audit3. Runtime. Add channel transfer feature to the content pallet.

shamil-gadelshin opened the pull request:
[Original issue](https://github.com/Joystream/joystream/issues/2816)

## Changes
- new extrinsic `update_channel_transfer_status`
- new extrinsic `accept_channel_transfer`
- added 'pending transfer` check barriers for multiple extrinsics (prevent extrinsics call for channels with pending transfer)

## Comments
1.  Parameters validation
    - existing check for the `new_collaborators` was used (validate_member_set)
    - no upper bound for `new_collaborators` was enforced similar to the existing code
    - no channel ownership or price checks were added
2. No price related logic was added

## Affected existing extrinsics
#### Modified extrinsics
- update_channel_censorship_status
- update_channel
- create_video
- update_video
- delete_video
- edit_post_text
- delete_post
- update_moderator_set
- claim_channel_reward
#### Skipped extrinsics
- delete_channel - doesn't make sense to prevent this operation
#### To be modified extrinsics ([separate issue](https://github.com/Joystream/joystream/issues/3433))
- start_nft_auction
- cancel_nft_auction
- cancel_offer
- cancel_buy_now
- make_bid
- cancel_open_auction_bid
- claim_won_english_auction
- pick_open_auction_winner
- offer_nft
- sling_nft_back
- accept_incoming_offer
- sell_nft
- buy_nft
- channel_owner_remark
- channel_collaborator_remark
- channel_moderator_remark
- nft_owner_remark

shamil-gadelshin committed:
runtime: content: Add extrinsic.

- add `update_channel_transfer_status` extrinsic.

shamil-gadelshin committed:
runtime: content: Add extrinsic.

- add `accept_channel_transfer` extrinsic with tests

shamil-gadelshin committed:
runtime: content: Bump version up to 4.1.0

shamil-gadelshin committed:
runtime: content: Add ‘pending transfer’ barrier.

- modify extrinsics: add ‘no pending transfers’ checks for several extrinsics.

shamil-gadelshin committed:
runtime: content: Add pending transfer barriers.

Add ‘pending transfer’ check barrier for extrinsics:
- claim_channel_reward
- edit_post_text
- delete_video
- update_moderator_set

shamil-gadelshin committed:
runtime: content: add channel transfer parameters validation.

shamil-gadelshin commented:
We should update the `audit3` branch with the code from `rhodes` branch prior to reviewing this PR.

bedeho commented on runtime-modules/content/src/lib.rs:
```diff
@@ -1948,6 +1949,38 @@ decl_module! {
+            ensure_actor_authorized_to_transfer_channel::<T>(origin, &actor, &channel.owner)?;
+
+            ensure!(
+                channel.transfer_status == ChannelTransferStatus::NoActiveTransfer,
```
This constraint is not required, I should be able to change terms of an ongoing transfer in any way, this is totally useful, e.g. reducing price or anything else.

shamil-gadelshin replied:
We forgot about the `cancel_transfer` extrinsic. But we could skip creating it if we allow `update_channel_transfer_status` to set any transfer status.

bedeho replied:
yes this was the idea.

bedeho commented on runtime-modules/content/src/lib.rs:
```diff
@@ -1948,6 +1949,38 @@ decl_module! {
+                Error::<T>::InvalidChannelTransferStatus
+            );
+
+            Self::validate_member_set(&new_status_params.new_collaborators)?;
```
I would not even bother with this, let people provide whatever ids they want, runtime does not need to babysit, and if we are doing this check other places I would say that is the mistake, and it should be removed in the interest of consistency, rather than adding here.

shamil-gadelshin replied:
Should, I remove all entries of the `validate_member_set` within this task?

- create_channel
https://github.com/Joystream/joystream/blob/b3d36b7ebf348cf083b79bbe904bce6ebb1ea341/runtime-modules/content/src/lib.rs#L355

- update_channel
https://github.com/Joystream/joystream/blob/b3d36b7ebf348cf083b79bbe904bce6ebb1ea341/runtime-modules/content/src/lib.rs#L461

- update_moderator_set
https://github.com/Joystream/joystream/blob/b3d36b7ebf348cf083b79bbe904bce6ebb1ea341/runtime-modules/content/src/lib.rs#L1235

bedeho replied:
yes I think so, thank you for chasing this down here in this PR. Some tests will probably break.

bedeho reviewed (changes requested):
Minor tweaks.

shamil-gadelshin committed:
runtime: content: Fix review comments.

shamil-gadelshin committed:
runtime: content: Remove member set validation.

Removed `validate_member_set` check for extrinsics:
- create_channel
- update_channel
- update_moderator_set

Initiated by review comment:
https://github.com/Joystream/joystream/pull/3432#discussion_r828832664

bedeho reviewed (approved):
Excellent work.

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-channel-transfer

# Conflicts:
#	runtime-modules/content/src/lib.rs

shamil-gadelshin committed:
runtime: content: Add test.

shamil-gadelshin committed:
runtime: content: Fix test.

shamil-gadelshin committed:
Merge branch 'audit3' into audit3-channel-transfer

# Conflicts:
#	runtime-modules/content/src/tests/channels.rs
#	runtime-modules/content/src/types.rs
